### ISSUE TITLE
ECDsa.Create(ECParameters) behavior is not consistent across platforms with a public/private key mismatch

### ISSUE THREAD

**Issue by bartonjs** [dotnet/runtime#27830]
Create an ECParameters for secp256r1 (ECCurve.NamedCurves.nistP256) with

`D = 9F9BD156374FB78F3D69EFF10DEF8C296EC4F03EACA42F4257130D0CE9316FCD`,
`Q.X = 6B17D1F2E12C4247F8BCE6E563A440F277037D812DEB33A0F4A13945D898C296`
`Q.Y = 4FE342E2FE1A7F9B8EE7EB4A7C0F9E162BCE33576B315ECECBB6406837BF51F5`

(That X/Y pair is for the curve point G, not the key point Q)

On Linux (via OpenSSL)

>  Unhandled Exception: Interop+Crypto+OpenSslCryptographicException: error:100B107B:elliptic curve routines:EC_KEY_check_key:invalid private key
at System.Security.Cryptography.ECOpenSsl.ImportParameters(ECParameters parameters)

On Windows, success (the Q value seems to be rederived from D).

Since it's probably hard to make Linux and macOS coerce Q, it's probably more practical to make Windows throw (theoretical process: NCryptImportKey, then NCryptExportKey and compare the target Q value, then dispose the new key handle and throw before replacing the existing key handle).

**Comment by vcsjones** [comment]
@bartonjs this has been on me for a while, sorry about that. I will pick this one back up again, but some new developments have happened. Since we can now re-compute Q from D on macOS and Linux, we can without a big lift (I don't think) make those match Windows' behavior. If creating with Q fails, but we have D, derive Q and try again.

This to me seems (probably) better since making Windows throw is more disruptive, and we tend to use Windows' behavior as a source of truth.

**Comment by bartonjs** [comment]
Yeah, I'm generally a fan of throwing, but since we already don't on netfx or corefx-win then "work everywhere" is an easier sell than "break some things" if both are viable.

**Comment by vcsjones** [comment]
(This is a brain dump of "stuff" just so future Kevin knows what present Kevin recently learned)

I picked this up (again, determined to make progress on it) and it seems that Windows' behavior is not entirely understood. It appears to re-derive Q... sometimes. This for example, will fail:

```c#
using ECDsa ecdsa = ECDsa.Create(ECCurve.NamedCurves.nistP256);
ECParameters ecParams = ecdsa.ExportParameters(true);
ecParams.Q.X.AsSpan().Fill(0xFF);
ecParams.Q.Y.AsSpan().Fill(0xFF);
ecdsa.ImportParameters(ecParams);
```

> Unhandled exception. Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: The requested operation is not supported.

My current thinking right now is that Windows will re-compute Q only if whatever Q it was supplied with is on the Curve (as we already know, CNG will also re-compute in the case of Q=(0,0) but it handles that differently)

Here is another example that "works" where the Q value is on the Curve, but G × D != Q:

```c#
using ECDsa ecdsa = ECDsa.Create(ECCurve.NamedCurves.nistP256);
ECParameters ecParams = ecdsa.ExportParameters(true);
ecdsa.GenerateKey(ECCurve.NamedCurves.nistP256);
ECParameters otherKey = ecdsa.ExportParameters(false);
ecParams.Q = otherKey.Q;
ecdsa.ImportParameters(ecParams);
```

Another interesting quirk is that CNG appears to export the original, bad, Q. Consider:

```c#
string key = @"MHcCAQEEIHChLC2xaEXtVv9oz8IaRys/BNfWhRv2NJ8tfVs0UrOKoAoGCCqGSM49AwEHoUQDQgAErZsBkBjHlngjE2f3bN0Jil7wEE/V1O4Py5WLjhP6TSnGBIG3xpZ3Kc2iXHA31di60AOVC1K8HVpivFKrvGIT9A==";
using ECDsa ecdsa = ECDsa.Create();
ecdsa.ImportECPrivateKey(Convert.FromBase64String(key), out _);
string exported = Convert.ToBase64String(ecdsa.ExportECPrivateKey());
Console.WriteLine(exported == key);
```

This will print "true". So as far as we can see, the key did not change when being washed through CNG.

But if you run the key through openssl, we can see it is invalid.

```shell
base64 --decode << 'EOF' | openssl ec -check -inform der -noout
MHcCAQEEIHChLC2xaEXtVv9oz8IaRys/BNfWhRv2NJ8tfVs0UrOKoAoGCCqGSM49AwEHoUQDQgAErZsBkB
jHlngjE2f3bN0Jil7wEE/V1O4Py5WLjhP6TSnGBIG3xpZ3Kc2iXHA31di60AOVC1K8HVpivFKrvGIT9A==
EOF
```

> EC Key Invalid!
> 140177670411584:error:1010207B:elliptic curve routines:ec_key_simple_check_key

What I am thinking then, if we want to match Windows.

1. OpenSSL we can get pretty close. We can distinguish between "point not on curve" and "This Q is not for that D".
     1. If `EC_KEY_set_public_key_affine_coordinates` fails when importing, we have it the case of "Q is not on the curve". We continue to let this fail to match Windows.
     2. We make it to `EC_KEY_check_key`  and that fails. This is what Windows is handling. If we fail here, and we have D, then we can re-compute Q.
     3. When "exporting", OpenSSL behavior will differ from Windows. OpenSSL will export the Q based on whatever it calculated from D. Windows will export whatever the caller supplied. If we really wanted OpenSSL to match Windows, we would have to stash the original Q for the lifetime of the key and personally I don't see a ton of value in that.


2. macOS does not let us make this distinction. The key either imports, or it doesn't. We can try again if the import fails without Q if we have D. MacOS would also have the same export characteristics as OpenSSL.

**Comment by bartonjs** [comment]
When Windows is in a mismatched Q state, what does the public key export say? (Imported Q or correct Q?)  Even if that's the "imported" Q we could... complicatedly... always double-import, once with Q as-was, once with Q=0, and then normalize the behavior across mismatches.

**Comment by vcsjones** [comment]
> what does the public key export say?

Everything is consistent. If I import with a wrong Q in `ImportECPrivateKey` then `ExportParameters(false)` and `ExportSubjectPublicKeyInfo` are the caller supplied Q, so "imported" everywhere.

>  we could... complicatedly... always double-import, once with Q as-was, once with Q=0

😅 This seems like a minor behavior deviation that doesn't justify the complexity, IMO. (Your house, your rules though). The end result would be consistent across platforms from signing / verification, even if you export it and import it it again. You just can't easily observe the Q correction in Windows.

**Comment by vcsjones** [comment]
On the macOS 11 front, the behavior here is... strange. No exception is raised anywhere but a simple signature round-trip does not work.

```c#
string key = @"MHcCAQEEIHChLC2xaEXtVv9oz8IaRys/BNfWhRv2NJ8tfVs0UrOKo
AoGCCqGSM49AwEHoUQDQgAErZsBkBjHlngjE2f3bN0Jil7wEE/V1O4Py5WLjhP6TSnGB
IG3xpZ3Kc2iXHA31di60AOVC1K8HVpivFKrvGIT9A==";
using ECDsa ecdsa = ECDsa.Create();
ecdsa.ImportECPrivateKey(Convert.FromBase64String(key), out _);
Console.WriteLine(Convert.ToBase64String(ecdsa.ExportECPrivateKey()));

byte[] data = new byte[16];
RandomNumberGenerator.Fill(data);

byte[] signature = ecdsa.SignData(data, HashAlgorithmName.SHA256);
bool valid = ecdsa.VerifyData(data, signature, HashAlgorithmName.SHA256);
Console.WriteLine(valid); // False!
```

This occurs in netcoreapp3.1 too, so there hasn't been a regression. I will dig to see if I can make any sense out of what is happening.

**Comment by vcsjones** [comment]
It looks like, because the Apple PAL has a handle for both the public key and the private key, we are tearing apart the public key and private key from ECPrivateKey in to `ECParameters`.

So we end up with a private key for D, and a public key for an unrelated Q. It seems then we are signing with one key and verifying with another.

The easiest fix is to ignore Q if we have D.
